src/compiler/program.ts

```diff
@@ -862,6 +865,24 @@ namespace ts {
                 return oldProgram.structureIsReused;
             }
 
+            // If a file has ceased to be missing, then we need to discard some of the old
+            // structure in order to pick it up.
+            // Caution: if the file has created and then deleted between since it was discovered to
+            // be missing, then the corresponding file watcher will have been closed and no new one
+            // will be created until we encounter a change that prevents complete structure reuse.
+            // During this interval, creation of the file will go unnoticed.  We expect this to be
+            // both rare and low-impact.
+            const missingFilePaths: Path[] = oldProgram.getMissingFilePaths() || emptyArray;
+            for (const missingFilePath of missingFilePaths) {
+                if (host.fileExists(missingFilePath)) {
+                    return oldProgram.structureIsReused = StructureIsReused.SafeModules;
+                }
+            }
+
+            for (const p of oldProgram.getMissingFilePaths()) {
+                filesByName.set(p, undefined);

```

Comment:
I'm happy to update the type, but I should mention that the map already had `undefined` values (this change just ensures that they persist across structure reuse).